### PR TITLE
Add keys array to StorageDictionaries

### PR DESCRIPTION
**Pull request by bearbin**
This PR depends on #265, so that one needs to be reviewed / merged first.

### Keys array updating

The keys array is updated in the same manner as the size of dynamic arrays - that is, on any access to a dictionary, the key is added to the keys array if it doesn't already exist. There is no provision for removing keys from the array as that would have to wait until optionals/a better defined way of initialising and deleteing storage objects.

### TODO;

 - [x] Add tests.
 - [x] Implement looping over dictionaries.

### CODE REVIEW

File: Tests/BehaviorTests/tests/array/test/test/test.js [hunk 1]
```diff
@@ -91,41 +112,50 @@ contract(config.contractName, function(accounts) {
...
-         
+
+  it("should return the correct size for non-empty dynamically sized arrays", async function() {
+      const instance = await Contract.deployed();
+      let t;
+
+      t = await instance.sizeUp3.call();
+      assert.equal(t.valueOf(), 2);
+  });
+<<<<<<< HEAD
```
Comment (DJRHails) [inline Tests/BehaviorTests/tests/array/test/test/test.js]: Here is your issue!

**Comment by bearbin** [comment]
For loops over dictionaries are now also possible, and I added a test for them.

File: Sources/AST/ASTVisitor.swift [hunk 2]
```diff
@@ -374,7 +374,16 @@ public struct ASTVisitor<Pass: ASTPass> {
       processResult.passContext.asLValue = false
     }
 
-    processResult.element.rhs = processResult.combining(visit(processResult.element.rhs, passContext: processResult.passContext))
+    switch passContext.environment!.type(of: processResult.element.lhs, enclosingType: passContext.enclosingTypeIdentifier!.name, scopeContext: passContext.scopeContext!) {
+    case .arrayType(_):
```
Comment (DJRHails) [inline Sources/AST/ASTVisitor.swift]: I'd recommend combining these cases using the comma notation for switch blocks

File: Sources/AST/Environment.swift [hunk 3]
```diff
@@ -350,6 +351,30 @@ public struct Environment {
...
+            fatalError()
+          }
+        case .fixedSizeArrayType(_):
+          if case .identifier(let identifier) = binaryExpression.rhs, identifier.name == "size" {
+            return .basicType(.int)
+          } else {
+            fatalError()
+          }
+        case .dictionaryType(_):
+          if case .identifier(let identifier) = binaryExpression.rhs, identifier.name == "size" {
```
Comment (DJRHails) [inline Sources/AST/Environment.swift]: Same as the prior discussion on the size. This needs to be generalised in future, but is sufficient for now

File: Sources/IRGen/Function/IULIAExpression.swift [hunk 4]
```diff
@@ -104,18 +104,42 @@ struct IULIAPropertyAccess {
...
+      }
+    case .arrayType(_):
+      if case .identifier(let identifier) = rhs, identifier.name == "size" {
+        rhsOffset = "0"
+      } else {
+        fatalError()
+      }
+    case .dictionaryType(_):
+      if case .identifier(let identifier) = rhs, identifier.name == "size" {
+        rhsOffset = "0" // TODO: check!
```
Comment (DJRHails) [inline Sources/IRGen/Function/IULIAExpression.swift]: Is this a leftover fragment or do you still need to add this check?

File: Tests/BehaviorTests/tests/bank/test/test/test.js [hunk 5]
```diff
@@ -147,8 +147,7 @@ contract(config.contractName, function(accounts) {
       const newBalance = web3.eth.getBalance(accounts[4]);
 
       // Estimate
-      const withdrawGasCost = registerGasCost / 2
-
+      const withdrawGasCost = 0
```
[the remark this answers is not in the document]
Comment (DJRHails) [reply Tests/BehaviorTests/tests/bank/test/test/test.js]: I think this is doing some basic gas estimation for the contract function - the changes made have caused the gas cost to change.

File: Sources/IRGen/Function/IULIAExpression.swift (hunk 4 above)
Comment (bearbin) [reply Sources/IRGen/Function/IULIAExpression.swift]: This is a leftover, I will remove it.

File: Tests/BehaviorTests/tests/bank/test/test/test.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (bearbin) [reply Tests/BehaviorTests/tests/bank/test/test/test.js]: I was confused as the error being given was that the calculated gas cost was now much higher than the actual, which conflicted with the expected result of my changes increasing the total amount of gas consumed. Reading it again it makes sense, as my changes will have increased the deposit cost without changing withdrawal cost that much, leading to the new withdrawal cost being an overestimate. I think it's safe to leave theis here as a new estimate - it's close enough.

**Pull request merged**